### ISSUE TITLE
Handle large number of connections

### ISSUE THREAD

**Issue by siskavera** [jborean93/smbprotocol#27]
We use the smbprotocol library to access a Samba share in our code. When the same client tries to open a large number of connections, the server appears to be overloaded and the operation fails, either due to insufficient credits or because of a broken pipe. Here are tracebacks for these two cases:

```
...
  File "/home/apache/adidas/custom/adidas/utilities/file_operation/smb_file_operation.py", line 91, in copy
    if smbclient.path.isdir(source_path, **self.common_kwargs):
  File "/usr/local/lib/python2.7/dist-packages/smbclient/path.py", line 110, in isdir
    return _stat_ismode(path, py_stat.S_ISDIR, True, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/smbclient/path.py", line 154, in _stat_ismode
    return check(stat(path, follow_symlinks=follow, **kwargs).st_mode)
  File "/usr/local/lib/python2.7/dist-packages/smbclient/_os.py", line 551, in stat
    query_info(transaction, FileAttributeTagInformation)
  File "/usr/local/lib/python2.7/dist-packages/smbclient/_io.py", line 202, in __exit__
    self.commit()
  File "/usr/local/lib/python2.7/dist-packages/smbclient/_io.py", line 228, in commit
    requests = self.raw.fd.connection.send_compound(send_msgs, sid, tid, related=True)
  File "/usr/local/lib/python2.7/dist-packages/smbprotocol/connection.py", line 981, in send_compound
    return self._send(messages, session_id=sid, tree_id=tid, related=related)
  File "/usr/local/lib/python2.7/dist-packages/smbprotocol/connection.py", line 763, in wrapped
    return func(self, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/smbprotocol/connection.py", line 1189, in _send
    % (credit_charge, credits_available))
SMBException: Request requires 1 credits but only 0 credits are available
```

```
...
  File "/home/apache/adidas/custom/adidas/utilities/file_operation/smb_file_operation.py", line 120, in exists
    return smbclient.path.exists(path, **self.common_kwargs)
  File "/usr/local/lib/python2.7/dist-packages/smbclient/path.py", line 27, in exists
    return _exists(path, False, True, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/smbclient/path.py", line 141, in _exists
    stat(path, follow_symlinks=follow_symlinks, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/smbclient/_os.py", line 551, in stat
    query_info(transaction, FileAttributeTagInformation)
  File "/usr/local/lib/python2.7/dist-packages/smbclient/_io.py", line 202, in __exit__
    self.commit()
  File "/usr/local/lib/python2.7/dist-packages/smbclient/_io.py", line 228, in commit
    requests = self.raw.fd.connection.send_compound(send_msgs, sid, tid, related=True)
  File "/usr/local/lib/python2.7/dist-packages/smbprotocol/connection.py", line 981, in send_compound
    return self._send(messages, session_id=sid, tree_id=tid, related=related)
  File "/usr/local/lib/python2.7/dist-packages/smbprotocol/connection.py", line 763, in wrapped
    return func(self, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/smbprotocol/connection.py", line 1244, in _send
    self.transport.send(send_data)
  File "/usr/local/lib/python2.7/dist-packages/smbprotocol/transport.py", line 67, in wrapped
    func(self, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/smbprotocol/transport.py", line 108, in send
    sent = self._sock.send(data)
error: [Errno 32] Broken pipe
```

Is there a way to handle such a case within the functionalities of this library?

**Comment by jborean93** [comment]
Are you able to clarify the following:

* Are you creating multiple connections to the same server, or just connecting to multiple servers
* Are these connections in the same Python process
* Are you using multithreading for concurrent access to the same server

I'm just trying to get a picture of how you generating the load so I can potentially test it out.

The error around 0 credits available is puzzling and potentially is a bug in the smbprotocol code as the server should always ensure the client has at least 1 credit available as per [Algorithm for the Granting of Credits](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-smb2/2e366edb-b006-47e7-aa94-ef6f71043ced)

> The server MUST ensure that the number of credits held by the client is never reduced to zero. If the condition occurs, there is no way for the client to send subsequent requests for more credits. 

If you are running requests in multiple threads then this might make sense as all the requests could be used up before a response was received from the server?

While I do want to get to the bottom of this I will have to say that the performance you get out of this implementation will never be as good as using some C library like Samba to mount the SMB share to the file system. My goal with this library is to achieve similar functionality at an acceptable speed.

**Comment by r2qokk** [comment]
I am facing a similar problem. I have the log shown like first one.

I would like to answer you questions first:
- I have multiple connections to the same server
- The connections are in the same process
- I am using multi-threading to access the same server
- I am good with the goal of this library

I have a django application that uses this library to connect to a samba server.
In each request if they needs to connect to the samba server, they have the flow:

1. Register session
2. See if a file exists by using `stat`
3. If it throws `SMBOSError` it means no such file, otherwise yes

The flow does not end here. However, it throws `SMBException` when just one or two connections presents.
Hope this helps. Cheers.

**Comment by aleliaert** [comment]
Am also encountering this issue - getting an exception about running out of credits.  On questions:
- Multiple connections to the same server
- Same python process
- Am trying to use multiple python threads, to write different files in parallel.

Note that our end goal is to get higher throughput to our SMB server, which is an instance of Amazon FSx for Windows.  So once past the "no creds" problem, might we then have a "want credits at a higher rate" issue?

Would having multiple SMB sessions open help with that then?  Is that possible up at the smbclient layer?

I also noticed that AWS advertises FSx as supporting SMB Multichannel, which is purported to allow one client to get full filesystem throughput.  Support for that would be neat - we have some big files passing through and maximizing single file write speed would be beneficial.

Am prepared to scale app horizontally, but a single python process is easier to code and operate.

Thanks and neat module - was very happy to find this when trying to avoid running my app on Windows.

**Comment by jborean93** [comment]
If you are running out of credits then the server is not able to handle the amount of resources required. There's not much the client can do about this as it's entirely governed on the server.

> I also noticed that AWS advertises FSx as supporting SMB Multichannel, which is purported to allow one client to get full filesystem throughput

This is not something I've looked into much but AFAIK it essentially opens a new session to the server using a new interface. This allows a client to have multiple network interfaces available to send data to the server increasing the amount of data that can be sent over the wire reducing time. Unfortunately this setup hasn't been implemented in smbprotocol, you can't even designate a specific interface for the client to use right now.

If speed is truly your goal then honestly you might be better off using Samba to mount the target to the file system and then read/write to that mount point in Python. You won't really get any faster than Samba as it's using C and is much more optimized than this library.

**Comment by aleliaert** [comment]
> you might be better off using Samba to mount the target to the file system
I tried that first - and ran into an issue with FUSE inside our container orchestration requiring some Docker capabilities that I couldn't get at the time.  Might revisit that.

Thanks for the quick response, and useful suggestion.

**Comment by jborean93** [comment]
Closing as this issue is quite old and based on the responses the problem lays with limited credits granted to the client by the server. There have been some updates in the credit calculations in https://github.com/jborean93/smbprotocol/pull/128 and https://github.com/jborean93/smbprotocol/pull/70 that might fix some of the bugs there but AFAIK the client should be calculating them correctly now.

While the code can theoretically run in multiple threads it's not something I've tested thoroughly so there is not guarantee it is bug free. If there is still a problem that can be replicated I can take a further look but for now I'm not sure there is anything else left to do here.
